orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -207,39 +280,274 @@ Status TrainingRunner::Run(IDataLoader* training_data_loader, IDataLoader* test_
+      auto it = opt_graph_outputs_.find(OptimizerOutputKey::GradientAccumulation);
+      ORT_RETURN_IF(it == opt_graph_outputs_.end(), "Gradient accumulation output is missing in the optimizer output");
       fetch_names.push_back(it->second);
     }
+
+    // Always execute event operators to avoid deadlock if pipeline is used.
+    // TODO: create a list of must-to-fetch tensors and pass it to all graph transformer.
+    if (params_.num_pipeline_stages) {
+        if (!pipeline_context_.forward_waited_output_name.empty()) {
+          fetch_names.push_back(pipeline_context_.forward_waited_output_name);
+        }
+        if (!pipeline_context_.forward_recorded_output_name.empty()) {
+          fetch_names.push_back(pipeline_context_.forward_recorded_output_name);
+        }
+        if (!pipeline_context_.backward_waited_output_name.empty()) {
+          fetch_names.push_back(pipeline_context_.backward_waited_output_name);
+        }
+        if (!pipeline_context_.backward_recorded_output_name.empty()) {
+          fetch_names.push_back(pipeline_context_.backward_recorded_output_name);
+        }
+    }
+  }
+
+  // We need to fetch at least one variable.
+  // If there is nothing to fetch, we fetch all model outputs.
+  if (fetch_names.empty()) {
+    fetch_names = allowed_fetch_names;
+  }
+
+  return Status::OK();
+}
+
+Status TrainingRunner::RunWithUpdate(VectorString& feed_names,
+                                     VectorString& fetch_names,
+                                     std::vector<MLValue>& feeds,
+                                     std::vector<MLValue>& fetches) {
+  pipeline_worker_pool_.join_all();
+
+  ORT_RETURN_IF_ERROR(session_.Run(RunOptions(),
+                                    feed_names,
+                                    feeds,
+                                    fetch_names,
+                                    &fetches));
+
+  if (loss_scaler_) {
+    auto it = std::find(fetch_names.begin(), fetch_names.end(), opt_graph_outputs_[OptimizerOutputKey::GradientAllIsFinite]);
+    if (it != fetch_names.end()) {
+      const size_t index = static_cast<size_t>(std::distance(fetch_names.begin(), it));
+      const Tensor& all_is_finite_t = fetches[index].Get<Tensor>();
+      const bool is_all_finite = *(all_is_finite_t.template Data<bool>());
+      loss_scaler_->UpdateLossScale(is_all_finite);
+    }
   }
 
-  VectorString fetch_grad_accumulator_output;
-  if (params_.gradient_accumulation_steps > 1) {
-    auto it = opt_graph_outputs_.find(OptimizerOutputKey::GradientAccumulation);
-    ORT_RETURN_IF(it == opt_graph_outputs_.end(), "Gradient accumulation output is missing in the optimizer output");
-    fetch_grad_accumulator_output.push_back(it->second);
+  // Assume that only the last pipeline stage can see loss, predicted value, and so on.
+  // Thus, the error function should only be called when we are at the last stage.
+  if (params_.num_pipeline_stages == 1 ||
```

Comment:
>|| [](start = 39, length = 2)

is this operator priority intended? not missing parenthesis?

Reply:
You're right. Thanks! It will be changed to
```
  const bool stage_can_see_loss = params_.pipeline_parallel_size == 1 ||
    pipeline_context_.pipeline_stage_id == params_.pipeline_parallel_size - 1;
  if (stage_can_see_loss &&
      !params_.is_perf_test &&
      weight_update_step_count_ % params_.display_loss_steps == 0) {
```